icechunk/src/cli/interface.rs

```diff
@@ -0,0 +1,140 @@
+    cmd: Command,
+}
+
+#[derive(Debug, Subcommand)]
+enum Command {
+    #[command(subcommand)]
+    Repo(RepoCommand),
+    #[command(subcommand)]
+    Snapshot(SnapshotCommand),
+}
+
+#[derive(Debug, Subcommand)]
+enum RepoCommand {
+    #[clap(name = "create")]
+    Create(CreateCommand),
+}
+
+#[derive(Debug, Subcommand)]
+enum SnapshotCommand {
+    #[clap(name = "list")]
+    List(ListCommand),
+}
+
+#[derive(Debug, Args)]
+struct CreateCommand {
+    repo: RepositoryAlias,
+}
+
+#[derive(Debug, Args)]
+struct ListCommand {
+    repo: RepositoryAlias,
+}
+
+fn load_repositories() -> Result<Repositories> {
+    let path = PathBuf::from("default.yaml");
+    let file = std::fs::File::open(path).context("❌ Failed to open config")?;
+    let deserialized: Repositories = serde_yaml_ng::from_reader(file)?;
+    Ok(deserialized)
+}
+
+async fn get_storage(
+    repository_definition: &RepositoryDefinition,
+) -> Result<Arc<dyn Storage + Send + Sync>> {
+    match repository_definition {
+        RepositoryDefinition::LocalFileSystem { path, .. } => {
+            let storage = new_local_filesystem_storage(&path)
+                .await
+                .context(format!("❌ Failed to create storage at {:?}", path))?;
+            Ok(storage)
+        }
+        RepositoryDefinition::S3 {
+            location, object_store_config, credentials, ..
+        } => {
+            let storage = new_s3_storage(
+                object_store_config.clone(),
+                location.bucket.clone(),
+                Some(location.prefix.clone()),
+                Some(crate::config::S3Credentials::from(credentials.clone())),
+            )?;
+            Ok(storage)
+        }
+    }
+}
+
+async fn repo_create(init_cmd: CreateCommand) -> Result<()> {
+    let repos = load_repositories()?;
+    let repo = repos.repos.get(&init_cmd.repo).context("❌ Repository not found")?;
+    let storage = get_storage(repo).await?;
+
+    let config = Some(repo.get_config().clone());
+
+    Repository::create(config, Arc::clone(&storage), HashMap::new())
+        .await
+        .context(format!("❌ Failed to create repository {:?}", init_cmd.repo))?;
+
+    println!("✅ Created repository {:?}", init_cmd.repo);
+
+    Ok(())
+}
+
+async fn snapshot_list(list_cmd: ListCommand) -> Result<()> {
+    let repos = load_repositories()?;
+    let repo = repos.repos.get(&list_cmd.repo).context("❌ Repository not found")?;
+    let storage = get_storage(repo).await?;
+    let config = Some(repo.get_config().clone());
+
+    let repository = Repository::open(config, Arc::clone(&storage), HashMap::new())
+        .await
+        .context(format!("❌ Failed to open repository {:?}", list_cmd.repo))?;
+
+    let snapshot_id =
+        repository.lookup_branch("main").await.context("❌ Failed to lookup branch")?;
+    let ancestry = repository.ancestry(&VersionInfo::SnapshotId(snapshot_id)).await?;
+    pin!(ancestry);
+    while let Some(snapshot_info) = ancestry.next().await {
```

Comment:
navigating ancestry in Icechunk is (currently) very slow. I'd recommend putting a limit of say 10 parents, and then let users override that default.

You may want to look at the `TryStreamExt` trait for useful methods in `Stream`, so you don't have to code the explicit loop. There is `try_map`, `take` and all what you would expect from an iterator type.

Reply:
Is this along the lines of what you were suggesting? https://github.com/earth-mover/icechunk/pull/716/commits/18d4d087105a1a40775da01fa76b598d26127265

Reply:
Great, and probably you no longer need to pin it

Reply:
https://github.com/earth-mover/icechunk/pull/716/commits/1c857e645bd4694a10a9024466e8716b48b5c844